[Enhancement]: Add SVM QoS Policy Activation

acch opened the pull request:
Adds ability to configure `qos_policy` parameter of SVMs.

Needs to be a dedicated resource, as creating a QoS Policy scoped to an SVM, and then activating it for the SVM, is a 2-step process. See [documentation](https://docs.netapp.com/us-en/ontap/performance-admin/set-throughput-ceiling-qos-task.html) for details.

Closes #301.

Acceptance tests pass:
```shell
$ TF_ACC=1 go test ./internal/provider/svm/svm_qos_policy_activation_resource_test.go -v
=== RUN   TestAccSvmQosPolicyActivationResource
--- PASS: TestAccSvmQosPolicyActivationResource (2.96s)
PASS
ok      command-line-arguments  2.967s
```

Example Terraform Configuration WIP:
```terraform
resource "netapp-ontap_svm_qos_policy_activation" "example" { 
  svm = {
    name = "svm01"
  }
  qos_policy = {
    name = "performance_svm01"
  }
}
```

acch committed:
feat: add `svm_qos_policy_activation` resource with numeric IDs

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: allow referencing SVM / QoS Policy by name

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add resource import function

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
chore: cleanup example

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add acceptance tests for `svm_qos_policy_activation` resource

Signed-off-by: Achim Christ <achim.christ@sva.de>

carchi8py reviewed (commented):
So do you expect users to ever pass in the UUID vs name?

When we wrote the Ansible/Terraform modules/resource, we've already assume users would always use name, and we would treat UUID as more of a internal thing. As a user is having both name and UUID useful?

suhasbshekar commented on internal/provider/svm/svm_qos_policy_activation_resource.go:
```diff
@@ -0,0 +1,407 @@
+						Optional:            true,
+						Computed:            true,
+					},
+					"id": schema.StringAttribute{
```
We try to use name all the time, as UUID is not something user friendly, and name itself can handle all functionality, so we have used name in all our Ansible/Terrafrom modules/resources. Unless you have an explicit case, let us know

acch commented:
There is an advantage of referencing dependent resources by their UUID: the ID is computed by ONTAP API, and hence it's "unknown" during plan-phase. Terraform uses this information when building its dependency graph, and implicitly understands that the dependent resources (e.g. `svm` or `qos_policy`) must be created *first*, before the `svm_qos_policy_activation` resource.

When referencing resources by their name, Terraform has no indication that one resource depends on the existence of another, and thus needs to be created first. It would attempt to create all resources in parallel (race condition). To prevent this, one would need to add the `depends_on` meta-argument.

So the following code...

```terraform
resource "netapp-ontap_qos_policy" "this" {
  name = "test"
}
resource "netapp-ontap_svm_qos_policy_activation" "this" { 
  qos_policy = {
    id = netapp-ontap_qos_policy.this.id
  }
}
```

...is equivalent to the following code:

```terraform
resource "netapp-ontap_qos_policy" "this" {
  name = "test"
}
resource "netapp-ontap_svm_qos_policy_activation" "this" { 
  qos_policy = {
    name = "test"
  }
  depends_on = [ netapp-ontap_qos_policy.this ]
}
```

Hence, I'd suggest to keep the UUID as an optional attribute - do you agree?

acch committed:
feat: add `svm_qos_policy_activation` documentation

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
I've manually added the documentation pages in the `docs/` directory. Please let me know what else is missing.

acch commented:
@carchi8py Are we good to merge this PR?

Please let me know what I need to do exactly to have the PRs added to the provider. My understanding is that I've addressed all requested changes... please let me know what's missing.

Many thanks!

suhasbshekar commented:
Can you update CHANGELOG.md file?

Also your branch is behind the integration/main, we added few more checks to the pipeline as it was missing here, since your PR is from a fork of this repo, so could you pull the new changes into your branch and update the PR?

Just for our reference, since we have disabled ACC tests as our testing profiles are down, acc tests are passing for this PR when tested locally.

acch committed:
chore: update changelog

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
@suhasbshekar Thanks for your feedback.

- I've added an entry to CHANGELOG.md
- I've rebased my branch to integration/main (once again)

Acc tests are passing on my local test system (ONTAP Select running 9.15.1P4):

```shell
$ date && TF_ACC=1 go test ./internal/provider/svm/svm_qos_policy_activation_resource_test.go -v && date
Tue Feb  4 10:18:03 UTC 2025
=== RUN   TestAccSvmQosPolicyActivationResource
--- PASS: TestAccSvmQosPolicyActivationResource (2.41s)
PASS
ok      command-line-arguments  2.421s
Tue Feb  4 10:18:06 UTC 2025
```

Please let me know what else is needed.
Many thanks!

suhasbshekar commented on internal/provider/svm/svm_qos_policy_activation_resource_test.go:
```diff
@@ -0,0 +1,92 @@
+
+	ntest "github.com/netapp/terraform-provider-netapp-ontap/internal/provider"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
```
can you replace this with "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"

acch replied:
Sure - done. Sorry, must have missed that.

acch committed:
fix: correct 'resource' package import

Signed-off-by: Achim Christ <achim.christ@sva.de>

suhasbshekar commented:
Please resolve the conflicts.

suhasbshekar commented:
Please resolve the conflicts.

acch commented:
@suhasbshekar I've rebased the PR once again.

Can you please let me know when you'll be able to merge these changes?

acch committed:
fix: add version check

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
fix: correct aggregates type

Signed-off-by: Achim Christ <achim.christ@sva.de>

suhasbshekar commented:
conflicts

suhasbshekar commented:
Also can you fix  the lint issue which was caused from merging the previous PR?
`Error: internal/provider/networking/network_ip_interface_resource_test.go:15:7: idRegex redeclared in this block
Error: 	internal/provider/networking/network_ethernet_port_resource_test.go:15:7: other declaration of idRegex (compile)
Error: Process completed with exit code 1.`

Please rebase once again to get all changes which was merged earlier and can go through the pipeline, so that something like above will not be missed.

acch committed:
fix: make local test const names unique

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
@suhasbshekar Conflicts resolved and linter errors fixed.

chuyich commented on internal/interfaces/svm.go:
```diff
@@ -24,15 +24,17 @@ type SvmDataModelONTAP struct {
 	Language       string              `mapstructure:"language,omitempty"`
 	MaxVolumes     string              `mapstructure:"max_volumes,omitempty"`
 	Name           string              `mapstructure:"name,omitempty"`
+	QoSPolicy      QoSPolicy           `mapstructure:"qos_policy"`
```
if both parameters in the QoSPolicy are omitempty, why QoSPolicy is not omitempty?

acch replied:
@chuyich Based on my testing, `omitempty` does not have any effect on attributes of type struct (such as `Ipspace` or `QoSPolicy`). It removes strings and numbers from the request body, but not structs.
